src/itkOMEZarrNGFFImageIO.cxx

```diff
@@ -433,41 +483,46 @@ OMEZarrNGFFImageIO::ReadImageInformation()
 
 // We call tensorstoreToITKComponentType for each type.
 // Hopefully compiler will optimize it away via constant propagation and inlining.
-#define ELEMENT_READ(typeName)                                                                        \
+#define READ_ELEMENT_IF(typeName)                                                                     \
   else if (tensorstoreToITKComponentType(tensorstore::dtype_v<typeName>) == this->GetComponentType()) \
   {                                                                                                   \
-    auto * p = reinterpret_cast<typeName *>(buffer);                                                  \
-    auto   arr = tensorstore::Array(p, store.domain().shape(), tensorstore::c_order);                 \
-    tensorstore::Read(store, tensorstore::UnownedToShared(arr)).value();                              \
+    ReadFromStore<typeName>(store, m_IORegion, reinterpret_cast<typeName *>(buffer));                 \
   }
 
 void
 OMEZarrNGFFImageIO::Read(void * buffer)
 {
+  // Use a proxy measure (voxel count) to determine whether we are reading
+  // the entire image or an image subregion.
   // This comparison needs to be done carefully, we can compare 3D and 6D regions
-  if (this->GetLargestRegion().GetNumberOfPixels() != m_IORegion.GetNumberOfPixels())
+  if (this->GetLargestRegion().GetNumberOfPixels() == m_IORegion.GetNumberOfPixels())
```

Comment:
Will this not cause an exception if entire image is read at once?

Reply:
No, this block is updated to simply verify that the size of the store matches the requested IO region. Existing CTests reading the entire image pass on my local system, will watch CI to verify.